Svc/FileDownlink/FileDownlink.cpp

```diff
@@ -49,160 +58,268 @@ namespace Svc {
+      //If the last packet has been sent (and is returning now) then finish the file
+	  if (this->lastCompletedType == Fw::FilePacket::T_END ||
+          this->lastCompletedType == Fw::FilePacket::T_CANCEL) {
+          finishHelper(this->lastCompletedType == Fw::FilePacket::T_CANCEL);
+          return;
+      }
+      //If waiting and a buffer is in-bound, then switch to downlink mode
+      else if (this->mode.get() == Mode::WAIT) {
+          this->mode.set(Mode::DOWNLINK);
+      }
+
+      this->downlinkPacket();
+  }
 
   // ----------------------------------------------------------------------
   // Command handler implementations 
   // ----------------------------------------------------------------------
 
   void FileDownlink ::
-    FileDownlink_SendFile_cmdHandler(
+    SendFile_cmdHandler(
         const FwOpcodeType opCode,
         const U32 cmdSeq,
         const Fw::CmdStringArg& sourceFileName,
         const Fw::CmdStringArg& destFileName
     )
   {
-
     Os::File::Status status;
 
-    if (this->mode.get() == Mode::CANCEL) {
-      Fw::LogStringArg sourceLogStringArg(sourceFileName);
-      Fw::LogStringArg destLogStringArg(destFileName);
-      this->log_ACTIVITY_HI_FileDownlink_DownlinkCanceled(
-          sourceLogStringArg,
-          destLogStringArg
-      );
-      this->mode.set(Mode::IDLE);
-      this->cmdResponse_out(opCode, cmdSeq, Fw::COMMAND_OK);
-      return;
+    // We are processing a file currently, reject command
+    if (this->mode.get() != Mode::IDLE) {
+        this->cmdResponse_out(opCode, cmdSeq, Fw::COMMAND_EXECUTION_ERROR);
+        return;
     }
 
-    this->mode.set(Mode::DOWNLINK);
+    // Store the current command info for completion later
+    this->curOpCode = opCode;
+    this->curCmdSeq = cmdSeq;
 
+    // Open file for downlink
     status = this->file.open(
         sourceFileName.toChar(),
         destFileName.toChar()
     );
-    if (status != Os::File::OP_OK) { 
+
+    // Reject command if error when opening file
+    if (status != Os::File::OP_OK) {
+      this->mode.set(Mode::IDLE);
       this->warnings.fileOpenError();
-      this->cmdResponse_out(opCode, cmdSeq, Fw::COMMAND_EXECUTION_ERROR);
-      return; 
+      if (FILEDOWNLINK_COMMAND_FAIL_ON_MISSING_FILE) {
+          this->cmdResponse_out(opCode, cmdSeq, Fw::COMMAND_EXECUTION_ERROR);
+      } else {
+          this->cmdResponse_out(opCode, cmdSeq, Fw::COMMAND_OK);
+      }
+      return;
     }
+    this->getBuffer(this->buffer, FILE_PACKET);
 
+    // Send file and switch to WAIT mode
     this->sendStartPacket();
+    this->mode.set(Mode::WAIT);
+    this->sequenceIndex = 1;
+    this->curTimer = 0;
+    this->byteOffset = 0;
+    this->endOffset = this->file.size;
+    this->lastCompletedType = Fw::FilePacket::T_START;
+    this->log_ACTIVITY_HI_SendStarted(this->file.size, this->file.sourceName, this->file.destName);
+  }
+
+  void FileDownlink ::
+    SendPartial_cmdHandler(
```

Comment:
Would it make sense to refactor SendPartial and SendFile into a single common function? It seems like there's a some duplicated logic.

Reply:
I agree, but I don't want that as this PR.  The existing code is as-is from the contributing project.